src/hotspot/share/prims/foreign_globals.cpp

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (c) 2020, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+#include "precompiled.hpp"
+#include "foreign_globals.hpp"
+#include "classfile/symbolTable.hpp"
+#include "memory/resourceArea.hpp"
+#include "runtime/fieldDescriptor.hpp"
+
+#define FOREIGN_ABI "jdk/internal/foreign/abi/"
+
+int ForeignGlobals::field_offset(InstanceKlass* cls, const char* fieldname, Symbol* sigsym) {
+  TempNewSymbol fieldnamesym = SymbolTable::probe(fieldname, (int)strlen(fieldname));
+  fieldDescriptor fd;
+  bool success = cls->find_field(fieldnamesym, sigsym, false, &fd);
+  assert(success, "Field not found");
+  return fd.offset();
+}
+
+InstanceKlass* ForeignGlobals::find_InstanceKlass(const char* name, TRAPS) {
+  TempNewSymbol sym = SymbolTable::probe(name, (int)strlen(name));
+  Klass* k = SystemDictionary::resolve_or_fail(sym, true, CHECK_NULL);
+  InstanceKlass* ik = InstanceKlass::cast(k);
+  ik->initialize(CHECK_NULL);
+  return ik;
+}
+
+const ForeignGlobals& ForeignGlobals::instance() {
+  static ForeignGlobals globals; // thread-safe lazy init-once (since C++11)
+  return globals;
+}
+
+const ABIDescriptor ForeignGlobals::parseABIDescriptor(jobject jabi) {
+  return instance().parseABIDescriptor_impl(jabi);
+}
+const BufferLayout ForeignGlobals::parseBufferLayout(jobject jlayout) {
+  return instance().parseBufferLayout_impl(jlayout);
+}
+
+ForeignGlobals::ForeignGlobals() {
+  TRAPS = Thread::current();
```

Comment:
Nit: You only need to introduce:
`Thread* THREAD`
if your code uses the CHECK macros. If you just want to pass the current thread as the callees "TRAPS" parameter you can pass any variable ie.
```
Thread* current = Thread::current();
...
InstanceKlass* k_ABI = find_InstanceKlass(FOREIGN_ABI "ABIDescriptor", current;
```

Reply:
Okay, will change it to be simpler. FWIW, I saw another place in our (Panama) code do this, so I subconsciously copied that pattern, thinking it was the norm. But, that code is the only place that uses that pattern as well.